### ISSUE TITLE
BattleArena v4.0 List of New Features

### ISSUE THREAD

**Issue by Europia79** [BattlePluginsDev/BattleArena#14]
This post is going to be a list of features that we want for BattleArena v4.0 (we have so many ideas that we want to implement that I feel we need a good way to keep track of them).

**BattleArena v4.0 Features:**
{

}

**Comment by Zach443** [comment]
That is a pretty good list to start with. I may add some things as I think of them.

**Comment by tara81** [comment]
Can u please fix the major bugs first =(

**Comment by tara81** [comment]
Like the scoreboards disapear all the time, their not even functional...

Players get trapped inside of battlearena sometimes when their not in the game nor queue and cannot join any more games until the last arena they were in is cancelled or a server reboot.

In custom gamemodes , there is teamkilling.  

Theres probably more but those are off the top of my head.

**Comment by Zach443** [comment]
@tara81 Scoreboards are in the list to be fixed, and I will be working on those. As for those other things you said, please create an issue on the appropriate repositories, that way we can see them in an organized way. Europia and I are really busy right now, and having things in an organized place will help us out a lot.

**Comment by Vervatwala** [comment]
Hey! Love the ideas so far (esp Multi rounds per match) 

Some Ideas I had were about Tourneys - choosing nteams/ team size seems to be broken (currently works the best  1v1s only) , but also itd be cool if we could have it choose a particular arena within the competition. 
Atm Tourneys seems to be struggling with playing with more than nteams=2 and anything bigger than a team of 1. 

A request I have (not very important but useful I feel) is when auto joining a game to use the arenas defaults not its competitions defaults. 
An example would be lets say there are 2 Skirmish arenas (one team based (one) and one FFA based (two)) so atm if one types ./sk join one (or two) the match will run with what ever Ive set as SetArenaOption. 
However if ppl simply do ./Sk join to auto join a match then the plugin is using the config defaults rather than its arena option. 
(Hope all tht above made sense if not Ill do my best to explain over TS/ IRC) 

Also a feature I think would be cool is to have 1 arena  with multimodes. so the same arena could function as a CTF map and a TDM map. (Obv wouldn't work with things like spleef etc)

**Comment by Europia79** [comment]
> Can u please fix the major bugs first =(

Yes, this is a major principle in software development: That bugs should be fixed before new features are implemented.  But rarely do we get indepth, detailed bug reports that help replicate bugs... So as a necessary first step, we have to waste time trying to replicate the bug first.  

Considering the release of SpongeAPI and the fact that there are so many bugs in BattleArena, I felt that this was a good time for a complete re-write... Because frankly, it's extremely difficult to pin-point the cause of some of these bugs, and on top of that, other so-called "bugs" are simply a product of a mis-configured BattleArena... BattleArena has so many configuration options that it's possible to make a mistake and create a precieved "bug."

This is really Alkarin's project and code-base, if he were here, I'm sure it would be no problem at all for him to fix some of the bugs, because he is more intimately familiar with the code-base. For other developers, it takes a little more effort to track down the bugs. With a complete/partial re-write I would be more intimately familiar with the code-base (if a bug arises). At a very minimum, I think we'll need at least a partial re-write and heavy refactoring, and bug squashing.

But in order to do this: We will need a road-map (software specification) of all the features that we want. 

> Also a feature I think would be cool is to have 1 arena with multimodes. so the same arena could function as a CTF map and a TDM map. (Obv wouldn't work with things like spleef etc)

Yeah, I think someone else proposed this also... i'll add it to the spec.

> A request I have (not very important but useful I feel) is when auto joining a game to use the arenas defaults not its competitions defaults. 
> An example would be lets say there are 2 Skirmish arenas (one team based (one) and one FFA based (two)) so atm if one types ./sk join one (or two) the match will run with what ever Ive set as SetArenaOption. 
> However if ppl simply do ./Sk join to auto join a match then the plugin is using the config defaults rather than its arena option. - Vervatwala

oh wow, thanks for the info.

Yes, I would definitely love to talk over TS / IRC / Skype.

**Comment by Zach443** [comment]
+1 to everything Europia said. BattleArena is a very large project, and it is difficult for us to adapt and track down bugs. Some of the things we have listed here may not seem important, but many of them will help us fix bugs at a much faster rate.  

@Vervatwala I love those ideas, particularly the one about multiple game modes per arena. I've been thinking that would be cool for a while now.

**Comment by tara81** [comment]
sorry, i reported a major bug last month and its still there, it's hard to
cope with

**Comment by tara81** [comment]
wasn't trying to be rude, i offered zach some money to fix it but I guess
its a hard one to fix, the one with players getting stuck in matches

**Comment by Zach443** [comment]
@tara81 Sorry about the seeming lack of urgency.. I've been super busy with school, and I know Euro has work. Overall we have been really bogged down lately, but we are coming up with some more time and getting the ball rolling a bit. Sorry for the wait, and thanks for staying with us.

**Comment by Europia79** [comment]
@tara81 I didn't think you were being rude.

The reason why that bug isn't getting fixed is because I cannot replicate it.... Meaning, it doesn't exist on my end.

You will have to jump on Skype, invite me to your server, and show me the bug in action and show me your configs.

**Comment by tara81** [comment]
its a random bug and hard to replicate, it happens to a player about once
every hour on my server sometimes sooner

**Comment by tara81** [comment]
It almost always involves a player leaving a minigame i think with the
/arena leave command

**Comment by Vervatwala** [comment]
Hey! Im sorry I just have one more thing I think would be cool: 

For Item Spawns (Wiki: http://wiki.battleplugins.org/BattleArena/Spawns ) 
A feature where the item or chest spawn has a particle effect around it similar to what the Flag in Arena CTF has. it makes finding a weapon or something easier to find for larger maps I think

With the current version on BA I'll try do some tests and see if there are any bugs with the current system and post bugs separately if any - Thank u

**Comment by tara81** [comment]
Request : More game victory conditions.  Eg: holding an area(arathi basin style), breaking a certain block (tower rush style)

**Comment by Europia79** [comment]
Correct me if I'm wrong... but Arathi Basin is pretty much Call of Duty: Domination... except, in Dom, kills give you points, and AB, kills give zero points ? i had plans to eventually do this type of game-mode... unfortunately, Search-N-Destroy was the most requested... so it got done first. 

For Tower-Rush-style... can you elaborate the rules of the game more ? Also, keep in mind that the TIME that it takes for a player to destroy a block (in default Minecraft) is NOT constant... Players with a good connection will be able to break blocks much faster than players with a horrible connection... So this will probably have to be corrected for in order to make the game fair.

**Comment by Zach443** [comment]
@tara81 For the tower rush style victory condition, I have been thinking about something like this for a while. I was thinking of implementing it as an extension to BattleArena (i.e. seperate plugin) rather than a VC, but I will think about it :smiley:

**Comment by tara81** [comment]
well alkarinv made me a tower rush plugin for $40 , I dont mind if you
update it and make it public.  It currently is broken however , since 1.8.

Here is the dropbox:
https://www.dropbox.com/s/34rn9xb9f9v3v2e/ArenaTowerRush.jar?dl=1

**Comment by johngonole** [comment]
I'd like to see the capture the flag extension and the paintball plugin extension combined so that players can have a capture the flag paintball match.   That would be really cool.

**Comment by Europia79** [comment]
> I'd like to see the capture the flag extension and the paintball plugin extension combined so that players can have a capture the flag paintball match. That would be really cool.

https://github.com/BattlePlugins/BattleArena/commit/7b3bebd32a6a73e701797e8138f9e8baf53f9a43

Hell yes, I wanted the same thing too ! That's why i fixed the Mod-Loader in the above commit.

http://wiki.battleplugins.org/BattleArena/Modules

First, you "install" all your mods to ```plugins/BattleArena/modules/```

Installation is just what's available for use.

Then, you have to specifically declare what mod(s) you want active for what arena(s). For example, in your case, you'd edit ```/plugins/ArenaCTF/CaptureTheFlagConfig.yml``` and use any valid yaml syntax to list all the mods you want activated during CTF matches. Then bam! You have CTF with Paintballs !

**Comment by RednedEpic** [comment]
Please refer to #106 for up-to-date information about what is to come.
